libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -202,8 +342,24 @@ static bool validateTransitionAck( MQTTPublishState_t currentState,
             break;
 
         case MQTTPubCompPending:
-            /* Outgoing publish, Qos 2. */
-            isValid = ( newState == MQTTPublishDone ) ? true : false;
+
+            /* Outgoing publish, Qos 2.
+             * There are 2 valid transitions possible.
+             * 1. MQTTPubCompPending -> MQTTPublishDone : A PUBCOMP is received.
+             *    This marks the complete state transistion for the publish packet.
+             *    This is the normal state transition without any connection
+             *    interuptions.
+             * 2. MQTTPubCompPending -> MQTTPubCompPending : Resending a PUBREL for
+             *    packets in state #MQTTPubCompPending can result in this
+             *    transition to the same state.
+             *    This can happen in the below state transition.
+             *    1. A TCP connection failure happened before receiving a PUBCOMP
+             *       for an outgoing PUBREL.
+             *    2. Reestablish MQTT session.
+             *    3. Resend the un-acked PUBREL will result in this transition
```

Comment:
Nit: (grammar), I'd change the tenses in each of these steps to agree with each other:
```suggestion
             *    1. A TCP connection failure happens before receiving a PUBCOMP
             *       for an outgoing PUBREL.
             *    2. MQTT session is reestablished.
             *    3. Resending the un-acked PUBREL results in this transition
```